src/compiler/parser.ts

```diff
@@ -2247,6 +2247,13 @@ namespace ts {
                 property.name = name;
                 property.questionToken = questionToken;
                 property.type = parseTypeAnnotation();
+
+                // Although object type properties cannot not have initializers, we attempt to parse an initializer
+                // so we can report that an object type property cannot have an initializer.
+                if (token === SyntaxKind.EqualsToken && lookAhead(() => parseNonParameterInitializer()) !== undefined) {

```

Comment:
You can just pass `parseNonParameterInitializer` to lookahead. Testing truthiness is fine here, and you'll avoid allocating a closure.